### ISSUE TITLE
Dynamic alternate urls

### ISSUE THREAD

**Issue by 50l3r** [nuxt-modules/i18n#2100]
Hi all,

I have a project in which certain routes are collected from a rest api. This is an example of a route:

ES: /camino-frances-sarria-grupo/
EN: /en/french-way-from-sarria-in-group/
IT: /it/cammino-francese-da-sarria-in-gruppo/

I have created a component that collects all the root routes `[slug].vue` since in the same root route there can be different types of content. I know it is not a good practice but I do not have the decisive power to change this.

The route as seen is the same since only the language code is placed before it but the `slug`changes depending on the language

```
<template>
    <component
        :is="component.component"
        v-for="(component, index) in components"
        v-once
        v-bind="component.payload"
        :key="index"
    />
</template>

<script setup lang="ts">
    import { defineAsyncComponent } from 'vue'
    import { useCms } from '@/modules/calltek/cms'

    const route = useRoute()
    const i18n = useI18n()
    const cms = useCms()
    const components: any[] = []

    try {
        const slug = route.params.slug.toString()
        const currentLang = i18n.locale.value

        const result = await cms.get(currentLang, ['ROUTE', 'PRODUCT'], slug)

        // alternate slugs
        const alernateUrls: Record<string, string> = {}
        result.data.slug.forEach((slug) => {
            alernateUrls[slug.code] = `/${slug.text}`
        })

        defineI18nRoute({
            paths: alernateUrls
        })

        components.push({
            component: defineAsyncComponent(
                () =>
                    import(`../partials/${result.type.toLowerCase()}/index.vue`)
            ),
            payload: {
                data: result.data,
                lang: result.locale
            }
        })
    } catch (error) {
        throw createError({
            statusCode: 404,
            statusMessage: 'Página no encontrada',
            fatal: true
        })
    }
</script>

<style scoped lang="scss"></style>
```

From what I see defineI18nRoute does not work with dynamic routes. How can I address this?

**Comment by 50l3r** [comment]
I try to do a hack like this:

```
route.params.slug_en = 'example-slug-en'
route.params.slug_it = 'example-slug-it'

defineI18nRoute({
    paths: {
        es: '/[slug]',
        en: '/[slug_en]',
        it: '/[slug_it]'
    }
})
```

But didn't work

**Comment by codeflorist** [comment]
@kazupon 

in my opinion it would be great to have a composable or compiler macro to define custom paths for ALL dynamic routes at once, that can be called, e.g. at the start of `app.vue`.
e.g.:

```js
defineI18nRoutes({
    paths: {
      about: {
        fr: '/a-propos',
      },
      me: {
        fr: '/je',
      },
      services: {
        fr: '/offres',
      },
      'services/advanced': {
        fr: '/offres/avancee',
      }
    }
})
```

this is practically a necessity with headless CMS, which are getting major traction and are a perfect fit for Nuxt.

this way, one could fetch all routes incl. their translated paths from the CMS and tell the i18n-module about it.

this would also enable linking to other pages using the `useLocalePath` cmposable, since i18n now has info about all the translated paths.

**Comment by FabianEllenberger** [comment]
I'm facing a similar issue with wrong alternate dynamic routes when using static generation with `nuxi generate` and using the strategy `prefix_except_default`.

The generated routes and translations seem fine but when trying to navigate via `useSwitchLocalePath()` from let's say the english page `en/test-page` to it's default route in german `test-seite`, nuxt tries to fetch the json payload that still includes the old locale `/en/test-seite/_payload.json` and fails because it doesn't exist. It should load `test-seite/_payload.json` without the `en` prefix.

On SSR this works fine, only facing this issue when generated staticly. It looks like the locale is not being updated fast enough only after nuxt is handling the GET for the payload.json file.

Thanks for any help in advance!

**Comment by 50l3r** [comment]
In my case I currently perform a predefined query to my backend in the paths /:slug and /:parent_slug/:slug

It happens that by hierarchy inherited from the old project i have different entities in those routes and I need to make a mandatory query to the backend so that it tells me which entity is the one I'm visiting and give me the data. The backend also returns me the routes of the alternative languages.

To probe all the routes at the beginning of the call would not be efficient in my case and at the moment I have a problem to generate the alternative links since as I commented previously I believe that it is not possible to define them in a dynamic way.

**Comment by Pezhvak** [comment]
This is an issue, how can we define alternative routes for dynamic parameter when loading data from an API?
Using `defineI18nRoutes` cannot solve this issue since it's compile time macro.

**Comment by 50l3r** [comment]
> This is an issue, how can we define alternative routes for dynamic parameter when loading data from an API? Using `defineI18nRoutes` cannot solve this issue since it's compile time macro.

I think that right now the only method is to delegate the route management to `nitro` or another `backend`.

**Comment by notflip** [comment]
> This is an issue, how can we define alternative routes for dynamic parameter when loading data from an API? Using `defineI18nRoutes` cannot solve this issue since it's compile time macro.

Did you ever find a solution to this? I'm getting my routes from an API and the alternate tags are wrong. I'm using

```
const setI18nParams = useSetI18nParams();
setI18nParams({
  en: {
   slug: data.en.slug
  },
  nl: {
    slug: data.nl.slug
})
```

To dynamically set the route slugs, but I suppose the alternate tags have already been set by the time this runs?

**Comment by BobbieGoede** [comment]
@notflip 
You're using `setI18nParams` correctly, please open a new issue with a minimal reproduction so we can take a closer look at your issue!

This issue should have been closed by https://github.com/nuxt-modules/i18n/pull/2580, so I'll be closing and locking it now (to prevent pinging the other participants in this and other very similar closed issues).
